### ISSUE TITLE
Case 2.10: don't reward denial-of-service behavior

### ISSUE THREAD

**Issue by wbudd** [crossbario/autobahn-testsuite#101]
[RingSocket](https://github.com/wbudd/ringsocket) follows this [RFC 6455 quote](https://tools.ietf.org/html/rfc6455#section-5.5.3) to the letter:

> If an endpoint receives a Ping frame and has not yet sent Pong frame(s) in response to previous Ping frame(s), the endpoint MAY elect to send a Pong frame for only the most recently processed Ping frame.

Sending multiple pings in rapid succession is much more likely to be a form of denial-of-service attack than to be the result of a valid practical scenario. Only responding to the last ping is a pragmatic way to mitigate that potential for abuse, so deviating from RFC 6455 the way autobahn-testsuite does in case 2.10 can IMO only be interpreted as a bug.

Below is what autobahn-testsuite reports as the reason for marking case 2.10's outcome as a **Fail**, whereas as I argue that the *Observed* data listed should be at least as valid/expected as the listed *Expected* data:

> **_Case Description_**
>
> Send 10 Pings with payload.
>
> **_Case Expectation_**
>
> Pongs for our Pings with all the payloads. Note: This is not required by the Spec .. but we check for this behaviour anyway. Clean close with normal code.
>
> **_Case Outcome_**
>
> Actual events differ from any expected.
>
> _Expected:_
> {'OK': [('pong', u'payload-0'), ('pong', u'payload-1'), ('pong', u'payload-2'), ('pong', u'payload-3'), ('pong', u'payload-4'), ('pong', u'payload-5'), ('pong', u'payload-6'), ('pong', u'payload-7'), ('pong', u'payload-8'), ('pong', u'payload-9')]}
>
> _Observed:_
> [('pong', u'payload-9')]

I note that you're aware of this spec deviation, but I believe it would be a good idea to change this behavior sometime.

<sub>PS: Autobahn-testsuite saved me lots of time testing WebSocket protocol details. Thank you providing this valuable tool!</sub>

**Comment by oberstet** [comment]
> I note that you're aware of this spec deviation, but I believe it would be a good idea to change this behavior sometime.

yes, strictly speaking, it is a testsuite bug.

> having said that: virtually all implementations of websocket I've seen reply to every ping .. so they skip the behavior the rfc would allow: only answer the last reply. then of course there is the practical question: as I am allowed to only answer the "latest ping" .. I can just never answer and pretend I would expect another ping yet. IMO, the rfc spec text is insufficient .. anyways;) it is unlikely that I find time to change the testsuite anytime soon ..

**Comment by wbudd** [comment]
I agree in the sense that I feel that RFC 6455 should have made an attempt at specifying in what cases a peer has sufficient justification for having "*not yet sent Pong frame(s) in response to previous Ping frame(s)*".

This is what happens in RingSocket's case: it tries to respond to the last ping frame (if any) of every "chop" (i.e., `read()` call). However, if that chop also contains one or more full or partial text/binary messages, the pong response is postponed until the last message has been fully parsed. Furthermore, if fragmentation causes parsing of said last message to carry over to the next chop, the postponed pong response may subsequently be cancelled in order to instead respond to any ping frame found in that next chop.

**Comment by oberstet** [comment]
curious, how do you read the spec? as previously discussed, the current text might render an implementation that never answers a ping valid. conforming to that is pointless, both for an implementation and a testsuite. the testsuite implementation renders any behavior that does not answer every ping immediately by a pong invalid.

**Comment by joffrey-bion** [comment]
As I said the spec doesn't actually specify what should be accepted by the endpoint that sends the pings.

The way I understand it for the peer that receives them is that it's ok to just send 1 pong for 3 pings. The idea is that you could cancel the job that is replying to previous pings when receiving a new one, and just reply to the last one.

So in my opinion, the test suite should accept receiving 1 pong after many pings on quick succession, without strictly expecting as many pongs as pings.

**Comment by oberstet** [comment]
> As I said the spec doesn't actually specify what should be accepted by the endpoint that sends the pings.

yes, which is why the testsuite tests 1 fully defined behavior (which is stricter than the spec) - but then wrongly flags any deviation from that with "red" (and this "red" instead of "yellow" flagging is a testsuite bug). agreed!

> it's ok to just send 1 pong for 3 pings.

"3" is arbitrary and undefined: I just wait for an infinite number of pings pretending to wait for a possibly next ping to be accumulated before sending any pong

> receiving 1 pong after many pings on quick succession

"quick" is undefined: I just wait for an infinite amount of time pretending to wait for the next ping in one succession to batch up before responding with any pong

---

anyways, personally I think the spec is unclear and the testsuite is slightly wrong, but that's "ok", as in, good enough;) but if you have a PR, we can change it! the red->yellow flagging I mean ..
